### PR TITLE
[CSBindings] Cache already computed type variable bindings

### PR DESCRIPTION
**Pull request by xedin**
To avoid re-computing same bindings over and over again,
let's try to keep a cache of active bindings and invalidate
them as needed.

Invalidate bindings conservatively when new constraints are
introduced (even if such constraints are not going to result
in new type bindings), and when type variables are assigned
fixed types or merged together.

Invalidation also makes sure to erase bindings of all adjacent
type variables related to equivalence class of the variables
in question.

This significantly speeds up use-cases which have a lot of type
variables such as collection literals with arbitrary nesting.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

File: validation-test/Sema/type_checker_perf/slow/rdar30596744_2.swift.gyb [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// RUN: %scale-test --invert-result --begin 1 --end 5 --step 1 --select incrementScopeCounter %s --expected-exit-code 1
+// RUN: %scale-test --invert-result --begin 1 --end 4 --step 1 --select incrementScopeCounter %s --expected-exit-code 1
```
Comment (slavapestov) [inline validation-test/Sema/type_checker_perf/slow/rdar30596744_2.swift.gyb]: Why did you change this?

File: lib/Sema/CSBindings.cpp [hunk 2]
```diff
@@ -25,22 +25,26 @@ Optional<ConstraintSystem::PotentialBindings>
...
   for (auto *typeVar : getTypeVariables()) {
     if (typeVar->getImpl().hasRepresentativeOrFixed())
       continue;
 
+    if (Bindings.count(typeVar) > 0)
+      continue;
+
     if (auto bindings = getPotentialBindings(typeVar))
-      cache.insert({typeVar, std::move(bindings)});
+      Bindings.insert({typeVar, std::move(bindings)});
```
Comment (slavapestov) [inline lib/Sema/CSBindings.cpp]: Do you want to assert that the insertion succeeded?

File: lib/Sema/CSSimplify.cpp [hunk 3]
```diff
@@ -1561,6 +1561,7 @@ ConstraintSystem::matchTypesBindTypeVar(
   }
 
   assignFixedType(typeVar, type);
+  invalidateBindings(typeVar);
```
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: Are there any other callers of assignFixedType() that need to do this?

File: lib/Sema/CSSimplify.cpp [hunk 4]
```diff
@@ -1688,6 +1689,7 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
 
         // Merge the equivalence classes corresponding to these two variables.
         mergeEquivalenceClasses(rep1, rep2);
+        invalidateBindings(rep1);
```
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: Maybe do this in mergeEquivalenceClasses()?

File: lib/Sema/CSBindings.cpp [hunk 5]
```diff
@@ -25,22 +25,26 @@ Optional<ConstraintSystem::PotentialBindings>
...
 
   // Now let's see if we could infer something for related type
   // variables based on other bindings.
   for (auto *typeVar : getTypeVariables()) {
-    auto cachedBindings = cache.find(typeVar);
-    if (cachedBindings == cache.end())
+    if (typeVar->getImpl().hasRepresentativeOrFixed())
+      continue;
+
+    auto cachedBindings = Bindings.find(typeVar);
```
Comment (slavapestov) [inline lib/Sema/CSBindings.cpp]: auto &?

**Review by slavapestov — COMMENTED** [review]
Some optional comments, but I don’t claim to understand this part of the type checker too well

File: validation-test/Sema/type_checker_perf/slow/rdar30596744_2.swift.gyb (hunk 1 above)
Comment (xedin) [reply validation-test/Sema/type_checker_perf/slow/rdar30596744_2.swift.gyb]: Cause scalw test can’t fit the curve with these changes unfortunately...

File: lib/Sema/CSSimplify.cpp (hunk 3 above)
Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: I am considering unifying both assign and merge paths, want to run perf test first before I spend more time on this...

File: lib/Sema/CSBindings.cpp (hunk 5 above)
Comment (xedin) [reply lib/Sema/CSBindings.cpp]: Ah yes, thank you! That’s an existing problem though

Comment (xedin) [reply lib/Sema/CSBindings.cpp]: Actually reference can't be used here because `find` return an iterator...

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
Failure was:
```
14:56:39 ld: warning: Could not find auto-linked library 'swiftMetal'
14:56:39 Undefined symbols for architecture x86_64:
14:56:39   "__swift_FORCE_LOAD_$_swiftMetal", referenced from:
14:56:39       __swift_FORCE_LOAD_$_swiftMetal_$_Vision in Vision.o
14:56:39      (maybe you meant: __swift_FORCE_LOAD_$_swiftMetal_$_Vision)
14:56:39 ld: symbol(s) not found for architecture x86_64
```

Let's try again...

**Comment by xedin** [comment]
@swift-ci please clean test source compatibility

**Comment by xedin** [comment]
@swift-ci please clean test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Review by rudkx — COMMENTED** [review]
So is the assumption here that having more types as potential bindings than is strictly necessary always safe? (which seems on the surface to be true)

If so it’s not clear that you actually need to invalidate bindings when merging type variables, but could instead merge bindings.

**Comment by xedin** [comment]
@rudkx Yes, that’s an assumption. I am just trying to take it one step at a time, the problem with updating bindings in-place is that we don’t currently record bindings which bind two type variables directly, so that has to be fixed before update in-place is possible...

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test

**Comment by rudkx** [comment]
SGTM!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please test macOS platform

**Comment by xedin** [comment]
@swift-ci please test source compatibility
